compiler/noirc_frontend/src/hir/comptime/interpreter/foreign.rs

```diff
@@ -1,40 +1,127 @@
     name: &str,
-    arguments: Vec<(Value, Location)>,
+    args: Vec<(Value, Location)>,
+    return_type: Type,
     location: Location,
 ) -> IResult<Value> {
+    use BlackBoxFunc::*;
+
     match name {
-        "poseidon2_permutation" => poseidon2_permutation(interner, arguments, location),
-        "keccakf1600" => keccakf1600(interner, arguments, location),
+        "aes128_encrypt" => aes128_encrypt(interner, args, location),
+        "bigint_from_le_bytes" => {
+            bigint_from_le_bytes(interner, bigint_solver, args, return_type, location)
+        }
+        "bigint_to_le_bytes" => bigint_to_le_bytes(bigint_solver, args, location),
+        "bigint_add" => bigint_op(bigint_solver, BigIntAdd, args, return_type, location),
+        "bigint_sub" => bigint_op(bigint_solver, BigIntSub, args, return_type, location),
+        "bigint_mul" => bigint_op(bigint_solver, BigIntMul, args, return_type, location),
+        "bigint_div" => bigint_op(bigint_solver, BigIntDiv, args, return_type, location),
+        "blake2s" => blake_hash(interner, args, location, acvm::blackbox_solver::blake2s),
+        "blake3" => blake_hash(interner, args, location, acvm::blackbox_solver::blake3),
+        "ecdsa_secp256k1" => ecdsa_secp256_verify(
+            interner,
+            args,
+            location,
+            acvm::blackbox_solver::ecdsa_secp256k1_verify,
+        ),
+        "ecdsa_secp256r1" => ecdsa_secp256_verify(
+            interner,
+            args,
+            location,
+            acvm::blackbox_solver::ecdsa_secp256r1_verify,
+        ),
+        "embedded_curve_add" => embedded_curve_add(args, location),
+        "multi_scalar_mul" => multi_scalar_mul(interner, args, location),
+        "poseidon2_permutation" => poseidon2_permutation(interner, args, location),
+        "keccakf1600" => keccakf1600(interner, args, location),
+        "range" => apply_range_constraint(args, location),
+        "sha256_compression" => sha256_compression(interner, args, location),
         _ => {
-            let item = format!("Comptime evaluation for builtin function {name}");
-            Err(InterpreterError::Unimplemented { item, location })
+            let explanation = match name {
+                "schnorr_verify" => Some("Schnorr verification will be removed.".into()),
+                "and" | "xor" => Some("It should be turned into a binary operation.".into()),
+                "recursive_aggregation" => Some("A proof cannot be verified at comptime.".into()),
+                _ => None,
+            };
+            if let Some(explanation) = explanation {
+                let item = format!("Attempting to evaluate foreign function '{name}'");
+                Err(InterpreterError::InvalidInComptimeContext { item, location, explanation })
+            } else {
+                let item = format!("Comptime evaluation for foreign function '{name}'");
+                Err(InterpreterError::Unimplemented { item, location })
+            }
```

Comment:
```suggestion
            let explanation = match name {
                "schnorr_verify" => "Schnorr verification will be removed.".into(),
                "and" | "xor" => "It should be turned into a binary operation.".into(),
                "recursive_aggregation" => "A proof cannot be verified at comptime.".into(),
                _ => {
                    let item = format!("Comptime evaluation for foreign function '{name}'");
                    return Err(InterpreterError::Unimplemented { item, location })
                }
            };

            let item = format!("Attempting to evaluate foreign function '{name}'");
            Err(InterpreterError::InvalidInComptimeContext { item, location, explanation })
```
style nit to avoid the additional pattern matching on the option